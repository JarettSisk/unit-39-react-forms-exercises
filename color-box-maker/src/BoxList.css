.BoxList {
    display: flex;
    flex-direction: column;
    align-items: center;
}

.BoxList-boxes {
  flex-wrap: wrap;
  border: 1px solid black;
  margin: 40px;
  display: flex;
}