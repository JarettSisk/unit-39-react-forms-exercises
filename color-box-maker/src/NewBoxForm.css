.NewBoxForm {
  text-align: center;
  display: flex;
  flex-direction: column;
  width: 300px;
}

.NewBoxForm-button {
  margin: 10px 0 10px 0;
}