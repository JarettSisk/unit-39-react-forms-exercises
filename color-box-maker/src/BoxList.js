import React, { useState } from "react";
import { v4 as uuid } from 'uuid';
import Box from "./Box";
import NewBoxForm from "./NewBoxForm";
import "./BoxList.css";

const BoxList = () => {
  const INITIAL_STATE = [
    {
      backgroundColor: "blue",
      height: "200px",
      width: "200px",
      id: uuid(),
      testId: 1
    },
    {
      backgroundColor: "green",
      height: "200px",
      width: "200px",
      id: uuid(),
      testId: 2
    }
  ];

  // id for testing
  const [testId, setTestId] = useState(3);

  const [boxes, setBoxes] = useState(INITIAL_STATE);

  // Remove box
  const remove = (boxId) => {
    // set boxes to new array only containing boxes that dont match the id of the one we want to remove
    setBoxes(boxes.filter(b => b.id !== boxId));
};

  // add box to our state
  const addBox = (box) => {
    const newBox = {...box, id: uuid(), testId: testId};
    setBoxes(boxes => [...boxes, newBox])
    // increment testId
    setTestId(testId + 1);
  }

  const renderBoxes = () => {
    return (
      boxes.map(box => {
        const {backgroundColor, height, width, id, testId} = box;
        return (
          <Box 
            testId={testId}
            key={id}
            id={id}
            backgroundColor={backgroundColor}
            height={height}
            width={width}
            remove={remove}
          />
        )
      })
    )
  }

  return (
    <div className="BoxList">
      <NewBoxForm addBox={addBox}/>
      <div className="BoxList-boxes">
        {renderBoxes()}
      </div>
    </div>
    
  )
}

export default BoxList;