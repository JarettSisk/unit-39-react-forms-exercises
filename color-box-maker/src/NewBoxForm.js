import React, { useState } from "react";
import "./NewBoxForm.css";

const NewBoxForm = ({ addBox }) => {
  const INITIAL_STATE = {
    backgroundColor: "",
    height: "",
    width: ""
  }
  
  // state for form
  const [formData, setFormData] = useState(INITIAL_STATE);

  // handle form submit
  const handleSubmit = (e) => {
    e.preventDefault();
    addBox(formData); //add the box in the parent state
    setFormData(INITIAL_STATE); // reset the form
  }

  // handle form change
  const handleChange = (e) => {
    const {name, value} = e.target;
    setFormData((data) => {
      return {
        ...data,
        [name]: value
      }
    })
  }


  // render form
  return (
    <form onSubmit={handleSubmit} className="NewBoxForm">
      <label htmlFor="backgroundColor">Background Color</label>
      <input
        type="text"
        id="backgroundColor"
        name="backgroundColor"
        value={formData.backgroundColor}
        onChange={handleChange}
        placeholder="pink"
      />

      <label htmlFor="height">Height</label>
      <input
        type="text"
        id="height"
        name="height"
        value={formData.height}
        onChange={handleChange}
        placeholder="100px"
      />

      <label htmlFor="width">Width</label>
      <input
        type="text"
        id="width"
        name="width"
        value={formData.width}
        onChange={handleChange}
        placeholder="200px"
      />

      <button className="NewBoxForm-button">Add new box!</button>
    </form>
  )   
}

export default NewBoxForm;