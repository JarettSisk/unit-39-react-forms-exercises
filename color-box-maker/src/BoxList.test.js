import React from "react";
import { render, fireEvent, screen } from "@testing-library/react";
import BoxList from "./BoxList";
import NewBoxForm from "./NewBoxForm";


// Smoke test
it("renders without crashing", function() {
    render(<BoxList />);
});
// Snapshot test
it("matches snapshot", () => {
    const { asFragment } = render(<BoxList/>);
    expect(asFragment()).toMatchSnapshot();
});

it("should add a new box", () => {
  const {queryByTestId, queryByLabelText, queryByText} = render(<BoxList />);
  const colorInput = queryByLabelText("Background Color");
  const heightInput = queryByLabelText("Height");
  const widthInput = queryByLabelText("Width");
  const submitNewBoxButton = queryByText("Add new box!");

  //1.  Expect there to be no pink box
  expect(queryByTestId("3")).not.toBeInTheDocument;

  // 2. set up inputs and submit to create new box
  fireEvent.change(colorInput, {target: {value: "pink"}});
  fireEvent.change(heightInput, {target: {value: "200px"}});
  fireEvent.change(widthInput, {target: {value: "200px"}});
  fireEvent.click(submitNewBoxButton);

  // 3. check that pink box exists
  const pinkBox = queryByTestId("3");
  expect(pinkBox).toBeInTheDocument;
  expect(pinkBox.style["_values"]["background-color"]).toBe("pink")

})