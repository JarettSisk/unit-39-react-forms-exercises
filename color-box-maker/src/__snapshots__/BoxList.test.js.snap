// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="BoxList"
  >
    <form
      class="NewBoxForm"
    >
      <label
        for="backgroundColor"
      >
        Background Color
      </label>
      <input
        id="backgroundColor"
        name="backgroundColor"
        placeholder="pink"
        type="text"
        value=""
      />
      <label
        for="height"
      >
        Height
      </label>
      <input
        id="height"
        name="height"
        placeholder="100px"
        type="text"
        value=""
      />
      <label
        for="width"
      >
        Width
      </label>
      <input
        id="width"
        name="width"
        placeholder="200px"
        type="text"
        value=""
      />
      <button
        class="NewBoxForm-button"
      >
        Add new box!
      </button>
    </form>
    <div
      class="BoxList-boxes"
    >
      <div
        class="Box"
        data-testid="1"
        style="background-color: blue; height: 200px; width: 200px;"
      >
        <button>
          X
        </button>
      </div>
      <div
        class="Box"
        data-testid="2"
        style="background-color: green; height: 200px; width: 200px;"
      >
        <button>
          X
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
